src/librustc_data_structures/obligation_forest/mod.rs

```diff
@@ -131,28 +146,32 @@ struct ObligationTreeId(usize);
 type ObligationTreeIdGenerator =
     ::std::iter::Map<::std::ops::RangeFrom<usize>, fn(usize) -> ObligationTreeId>;
 
+/// `usize` indices are used here and throughout this module, rather than
+/// `rustc_index::newtype_index!` indices, because this code is hot enough
+/// that the `u32`-to-`usize` conversions that would be required are
+/// significant, and space considerations are not important.
+type NodeIndex = usize;
+
 pub struct ObligationForest<O: ForestObligation> {
     /// The list of obligations. In between calls to `process_obligations`,
     /// this list only contains nodes in the `Pending` or `Waiting` state.
-    ///
-    /// `usize` indices are used here and throughout this module, rather than
-    /// `rustc_index::newtype_index!` indices, because this code is hot enough
-    /// that the `u32`-to-`usize` conversions that would be required are
-    /// significant, and space considerations are not important.
     nodes: Vec<Node<O>>,
 
-    /// A cache of predicates that have been successfully completed.
-    done_cache: FxHashSet<O::CacheKey>,
+    /// Nodes must be processed in the order that they were added which this list keeps track of.
```

Comment:
Could you expand on this? How does this list keep track of that? Why is it important?

Also, there are 3 lists here that seem to be suggestively grouped together -- presumably they store nodes in different states -- what is the order of things relative to one another in this list? Or is pending nodes a distinct list that overlaps the other two?